Allow to use a monotonic clock to prevent timing issues + more

kleisauke opened the pull request:
This PR:
- allows users to use a monotonic clock (`make USE_MONOTONIC_CLOCK=1`). `gettimeofday()` is affected by discontinuous jumps in the system time and should not be used to measure time intervals ([here's a good explanation](https://blog.habets.se/2010/09/gettimeofday-should-never-be-used-to-measure-time.html)).
- increases the precision of the theoretical arrival times from microseconds to nanoseconds. This is in line with the other implementations [<sup>[1]</sup>](https://github.com/brandur/redis-cell/blob/64d2321f57f49e27f5ce8d3e0ea5c8c3f4ddc9a8/src/cell/mod.rs#L172) [<sup>[2]</sup>](https://github.com/throttled/throttled/blob/608de1fa042e0b10dfa8a8298467bf5dc1e66367/rate.go#L213-L217).
- solves the issue mentioned in #1 by ensuring that `rater_limit` returns the TTL in milliseconds instead of seconds.
  `X-RateLimit-Remaining` before this PR:
  ```
  [kleisauke@PC-KAW ~]$ for run in {1..6}; do redis-cli RATER.LIMIT user 5 5 1 | sed -n '3~5p'; done
  5
  5
  5
  5
  5
  5
  ```
  After:
  ```
  [kleisauke@PC-KAW ~]$ for run in {1..6}; do redis-cli RATER.LIMIT user 5 5 1 | sed -n '3~5p'; done
  5
  4
  3
  2
  1
  0
  ```

Closes #1.

kleisauke commented:
@edufelipe Sorry for pinging you, but could you review this? If you want, I could split up this PR into several commits, which can be reviewed one-at-a-time.

Especially the last issue causes problems within our nginx rate limit module:
https://github.com/weserv/rate-limit-nginx-module/issues/1#issuecomment-506328650

edufelipe commented:
Hi @kleisauke!

Sorry about not being able to answer your pull request earlier. While I agree that having milliseconds instead of nanoseconds is an issue, I don't think that getting a monotonic clock here is the answer.

The main problem is the fact that the start point of the monotonic clock is unspecified. This means that in case we have replication in our redis instance and redis sentinel does a fail-over to other instance the reference clock would be severely wrong and rate-limits would be incorrectly applied.

We can mitigate the issue by explicitly disallowing the replication of rater-limit keys, or by just increasing the clock resolution without using the monotonic clock.

Perhaps that should be a setting when loading the module, so users can decide whats more important to them: the possibility of keeping existing limits during a fail-over or incorrect rate-limits in case of clock smearing.

edufelipe commented on ratelimit.c:
```diff
@@ -20,9 +20,46 @@
+
+/* miliseconds per second */
+#define MSEC_PER_SEC 1000LL
+
```
I think there should be another constant called `USEC_PER_NSEC` just so the code is more readable.

kleisauke replied:
The `gettimeofday()` call with the confusing logic is now removed. I don't think the `USEC_PER_NSEC` needs to be added anymore?

edufelipe commented on ratelimit.c:
```diff
@@ -20,9 +20,46 @@
+#ifndef CLOCK_REALTIME
+  struct timeval tv;
+  gettimeofday(&tv, 0);
+  return (((long long) tv.tv_sec) * NSEC_PER_SEC + tv.tv_usec) * MSEC_PER_SEC;
```
@kleisauke I think there might be a bug here. Shouldn't we use `USEC_PER_SEC` instead of `NSEC_PER_SEC` on the first multiplication?

kleisauke replied:
I've removed the `gettimeofday()` in favor of `clock_gettime()` with https://github.com/onsigntv/redis-rate-limiter/pull/4/commits/585e76592723594c2983768709b7856bec2bfe07, so this is no longer a issue.

kleisauke commented:
Hi @edufelipe,

Thanks for your review! I never thought about the failover process of Redis sentinel, so I made the monotonic clock opt-in with 585e765. By default the real time clock (`CLOCK_REALTIME`) is used, if users want to use the monotonic clock (`CLOCK_MONOTONIC`) it can be configured with:
```sh
make USE_MONOTONIC_CLOCK=1
```

I've also removed the `gettimeofday()` call with the confusing logic, it was there for backward compatible reasons because `clock_gettime()` is [not implemented in macOS before Sierra 10.12](https://stackoverflow.com/q/5167269/10952119).

Let me know if you want to support the older versions of macOS (for e.g. El Capitan), and if I need to update the `README.md` as well.

kleisauke commented:
Hi @edufelipe,

I just went through some old unmerged PRs, was there any particular reason why this wasn't merged? If you want, I could split-up this PR.

kleisauke committed:
Increase the precision of the theoretical arrival times

From microseconds to nanoseconds.

kleisauke committed:
Ensure TTL is returned in milliseconds instead of seconds

Resolves #1.

kleisauke committed:
Allow to use a monotonic clock to prevent timing issues

CLOCK_REALTIME is affected by discontinuous jumps in the system
time, including incremental adjustments performed by NTP.

kleisauke committed:
Fix broken link

kleisauke commented:
... updated to split-up the commits, let me know if you rather want separate PR's for this.

> I've also removed the `gettimeofday()` call with the confusing logic, it was there for backward compatible reasons because `clock_gettime()` is [not implemented in macOS before Sierra 10.12](https://stackoverflow.com/q/5167269/10952119).
> 
> Let me know if you want to support the older versions of macOS (for e.g. El Capitan)

I just noticed that the Makefile already defines 10.12 as minimum supported version, see:
https://github.com/onsigntv/redis-rate-limiter/blob/3d31d9b9b16528222d4650378b81c0569f736d21/Makefile#L10

So, I don't think it's necessary to implement support for these older versions of macOS.

kleisauke commented:
@edufelipe Friendly ping to help move this along. It currently prevents us from tagging a new release on the [weserv/rate-limit-nginx-module](https://github.com/weserv/rate-limit-nginx-module/) project (see issue https://github.com/weserv/rate-limit-nginx-module/issues/2).

edufelipe commented:
@kleisauke Sorry for the delay. It is now merged. Thanks for the contribution!
